### ISSUE TITLE
`globalThis` is only defined in simulation

### ISSUE THREAD

**Issue by Zomis** [exaV/screeps-kotlin-starter#18]
In the persistent world I got this in the console:

```
> globalThis
ReferenceError: globalThis is not defined
    at eval (eval at <anonymous> (_console1726734043946_0:1:46), <anonymous>:2:1)
    at _console1726734043946_0:1:46
    at _console1726734043946_0:1:60
    at Object.exports.evalCode (<runtime>:15348:63)
    at Object.exports.run (<runtime>:20993:41)
```

While in simulation I got this in the console:

```
> globalThis
[object DedicatedWorkerGlobalScope]
```

This makes the compiled JavaScript not work in the persistent world, only in the simulation.

So if I try to use my Kotlin code compiled to JavaScript in the persistent world, I get this:

```
ReferenceError: globalThis is not defined
    at Object.885 (main:1:22236)
    at n (main:1:25332)
    at main:1:25358
    at t.235.Object.defineProperty.value (main:1:117)
    at main:1:267
    at main:3:3
    at Object.exports.evalCode (<runtime>:15382:76)
    at Object.requireFn (<runtime>:21032:28)
    at Object.exports.run (<runtime>:20975:60)
```

If I turn off minification, I can see where exactly the error is:

```
var __WEBPACK_AMD_DEFINE_FACTORY__, __WEBPACK_AMD_DEFINE_ARRAY__, __WEBPACK_AMD_DEFINE_RESULT__;(function (root, factory) {
  if (true)
    !(__WEBPACK_AMD_DEFINE_ARRAY__ = [exports, __webpack_require__(36), __webpack_require__(711)], __WEBPACK_AMD_DEFINE_FACTORY__ = (factory),
		__WEBPACK_AMD_DEFINE_RESULT__ = (typeof __WEBPACK_AMD_DEFINE_FACTORY__ === 'function' ?
		(__WEBPACK_AMD_DEFINE_FACTORY__.apply(exports, __WEBPACK_AMD_DEFINE_ARRAY__)) : __WEBPACK_AMD_DEFINE_FACTORY__),
		__WEBPACK_AMD_DEFINE_RESULT__ !== undefined && (module.exports = __WEBPACK_AMD_DEFINE_RESULT__));
  else {}
}(globalThis, function (_, kotlin_kotlin, kotlin_io_github_exav_screeps_kotlin_types) { // <--- Error is on this line
  'use strict';
  //region block: imports
  var Unit_instance = kotlin_kotlin.$_$.c;
```

**Comment by Zomis** [comment]
A workaround for this problem:

Change in `build.gradle.kts`

    val modules = mutableMapOf<String, String>()
    modules["main"] = main.readText()

to

    val modules = mutableMapOf<String, String>()
    val fixGlobalThis = "if (!Game.rooms['sim']) globalThis = this;"
    modules["main"] = fixGlobalThis + main.readText()

**Comment by exaV** [comment]
Thanks @Zomis for the workaround! It probably helped a bunch of people because this was broken for more than a year. The issue is with how Kotlin bundles the js file. We need a "web" distribution to get minification through webpack but we don't actually publish for web. In 2.0.0 it was possible to fix this by adding `useCommonJs()` to the config but that breaks again with 2.2.

I found the `output.globalObject` config by accident which is exactly what we need here.

fixed with  8858877

**Comment by droidkfx** [comment]
That is a much better solution. I am very much a novice when it comes to the kotlin->js toolchain. TY

**Comment by droidkfx** [comment]
Also, just found out the hard way that, somewhere along the way, this change affects how variable and function names get minified. Deployed, and everything was like e5_x is undefined. Turns out for one reason or another, e5_x changed to e8_x. That structure happened to be stored in memory and chaos ensued 😆

**Comment by exaV** [comment]
Yes that was due to the change to es2015. Sorry about that. You can revert it in your own codebase by changing that line to `es5`. For new codebases I'd recommend using es2015. If your accidental "migration" is complete and your stuff works again you can stay on es2015 too.

**Comment by Zomis** [comment]
@droidkfx My suggestion is to only use external interfaces (that you define yourself) in memory and use `jsObject` to create them. That way you keep the structure and avoid any obfuscation issues.

**Comment by exaV** [comment]
I agree with Zomis about using only external interface. Additionally from Kotlin 2.0 onwards you don't need `jsObject` anymore. Instead you can annotate `external interface` with `@JsPlainObject` which will give you a constructor and a copy function.
